geps/gep-1742.md

```diff
@@ -364,7 +364,7 @@ sequenceDiagram
 
 Both timeout fields are string duration values as specified by
 [Golang time.ParseDuration](https://pkg.go.dev/time#ParseDuration) and MUST be >= 1ms
-or 0 to disable (no timeout).
+and <= 100y.
```

Comment:
I added some comments related to this in the API PR (https://github.com/kubernetes-sigs/gateway-api/pull/2013/files#r1245498559), but thinking about this more, I really don't see the value of specifying an upper limit.

The more I think about this, it seems that any value greater than a day or so, basically means no timeout. Certainly nobody would want to set a timeout to something like 5 years and really expect that the client and service will both be up and running in a paused state until it returns a value before the 5y deadline.

Unless I'm missing something, a user will want to configure the timeout to something relatively short, and any larger value really just mean "don't timeout", i.e., just hang until the client connection is terminated. So that's why I think the current design is right. User can set the timeout to some small value or 0 to disable timeouts. A user can also chose to set a pretty huge value which the implementation will likely support (up to the implementation limit of probably an int64 max value ns or ~300 years), but any specific number they chose that is large will work the same as 0 for all practical purposes. So documenting some arbitrary large max value is really meaningless IMO.

Reply:
One more related thought. If we want to say anything about max value it could be something like `the maximum value is implementation dependent, but MUST BE at least 1y`, or something like that.

That way the user can set it to any sensible value they'd like (>= 1ms) or 0 for infinite (don't timeout).

@robscott btw, you said:

> In Kubernetes APIs we always try to have a clear upper and lower bound for fields. This enables implementations to write tests that ensure they can support the min and max values for any field in the API.

How would an implementation test that a 10 year timeout works?

Reply:
>How would an implementation test that a 10 year timeout works?

Yeah, you're right, it's not really possible to write a test that waits this long, and any value that we can feasibly test is likely lower than at least some users would want. I think the real win here is that implementations can know that they support the full range of possible values allowed by the API just by understanding their dataplane.

> If we want to say anything about max value it could be something like the maximum value is implementation dependent, but MUST BE at least 1y, or something like that.

I think that's reasonable, but I'm struggling to think of any use case that would benefit from > 1y, so distinguishing beyond that is likely not worthwhile (again I could be wrong here). My strong bias is to start with tighter limits and only loosen them if a clearly defined need is presented.

>The more I think about this, it seems that any value greater than a day or so, basically means no timeout. Certainly nobody would want to set a timeout to something like 5 years and really expect that the client and service will both be up and running in a paused state until it returns a value before the 5y deadline.

Yeah I think that's probably accurate, I'm not sure what the right upper limit is here, but lower is probably better as a starting point.

>So documenting some arbitrary large max value is really meaningless IMO.

There's a very strong precedent in Kubernetes API review to prevent unbounded values. Even if we were to push forward without an upper limit here, I can't imagine this passing the final level of API review before v0.8.0. Both @thockin and @khenidak have strongly pushed to avoid any unbounded values in this API.

Reply:
Two things:

1) String-encoded durations is not the norm in kube APIs, and I think we should question whether that's the right way to represent durations.  The norm is `int32` with the units encoded in the name - e.g. `timeoutSeconds int32`, and the unit is almost always seconds.

2) It's not really unbounded if it is a float32 - there is a natural bound.

Reply:
That's a good point @thockin. I had thought of something like `timeoutSeconds int32` as well, it would certainly simplify the validation initially. A concern I'd have is that some users will want to be able to express more precise timeouts, for example GCP allows you to express timeouts in terms of nanoseconds, and most implementations support at least milliseconds as an input. That leaves us with a few options:

1. Only support one unit (probably seconds or milliseconds) - this matches the [approach of ingress-nginx](https://github.com/kubernetes/ingress-nginx/issues/6540) (seconds), [GCP BackendServices](https://cloud.google.com/load-balancing/docs/backend-service#timeout-setting) (seconds), [Kong](https://docs.konghq.com/kubernetes-ingress-controller/latest/references/annotations/#konghqcomconnect-timeout) (ms), and [HAProxy](https://github.com/haproxytech/kubernetes-ingress/blob/402c43c4698434d9d61b702772a2671ebf125987/documentation/custom-resources.md?plain=1#L11) (ms)
2. Support multiple units, combining the values (`timeout.seconds: 1` + `timeout.nanoseconds: 5000000` == `1.005s`). This roughly matches [Envoy](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto#envoy-v3-api-field-config-route-v3-routeaction-timeout) and [GCP URLMaps](https://cloud.google.com/compute/docs/reference/rest/v1/urlMaps#UrlMap.FIELDS.inlinedField_26) which both appear to use proto [duration](https://protobuf.dev/reference/protobuf/google.protobuf/#duration).
3. Support something like `metav1.Duration` as proposed by this GEP. This matches [Istio](https://istio.io/latest/docs/tasks/traffic-management/request-timeouts/), [Linkerd](https://linkerd.io/2.13/tasks/configuring-timeouts/), [Contour](https://projectcontour.io/docs/main/configuration/#timeout-configuration)

So with all of that said, we could start with 1 and leave room for 2. For example, that future could look like this:

GEP Today:
```yaml
timeouts:
  request: 10s
```

GEP Tomorrow:
```yaml
timeouts:
  request:
    seconds: 10
```

GEP Next Year:
```yaml
timeouts:
  request:
    seconds: 10
    nanoseconds: 500000
```

I'm not entirely sure that I prefer that over the more open-ended duration format, but it would be simpler to validate and more in line with other Kubernetes APIs. Does this make sense to you @thockin?

Reply:
@thockin @robscott 

> String-encoded durations is not the norm in kube APIs, and I think we should question whether that's the right way to represent durations. The norm is int32 with the units encoded in the name - e.g. timeoutSeconds int32, and the unit is almost always seconds.

I haven't worked much with kube APIs in the past, so I'm not sure I understand all the conventions and "rules", but given the use case here, where timeouts will typically be in milliseconds and sometimes seconds, the string serialization format used by time.Duration is really the best way to represent it IMO. I believe there is also an equivalent Rust duration type that also uses this same string format as well.

That said, I'm wondering if an exception from the usual kube rules makes sense, or is even possible, in this case?

Reply:
> and more in line with other Kubernetes APIs

With regards to timeouts, we use int32 exclusively in the main kube API.  I didn't even realize `metav1.Duration` exists, and it seemonly used in component APIs (i.e. config files).  Don't get me wrong, I wish we had used Duration everywhere - I do find it easier to deal with, but inconsistency is a pain.

Let me discuss a bit.

Reply:
OK this ended up being more complex than expected. @thockin started a [Slack thread](https://kubernetes.slack.com/archives/CJUQN3E4T/p1688063522998959) to discuss this more, that discussion will likely continue, but here's a high level summary:

1. There's not really any prior art for `metav1.Duration` in core Kubernetes APIs.
2. Open API v3 validation translates this to [xml dayTimeDuration](https://www.w3.org/TR/xmlschema11-2/#dayTimeDuration) ([source](https://github.com/OAI/OpenAPI-Specification/issues/607))
3. In Kubernetes, we'd be trying to parse this into a [Duration](https://pkg.go.dev/time#Duration), the formats look nearly identical, but there may be some differences in translation between types

I asked @howardjohn how Istio handles this - apparently they're using a plain `string` field for their timeouts and then using webhook validation to ensure the format lines up. That's theoretically an option for us, but does not work well if/when our webhook is not present.

I think we need more API reviewers to chime in here, but in general I'd prefer to move forward with a duration format as currently proposed if we can make it work. If not, I think our next best option would be to use something like `timeouts.request.milliseconds` as an `int32`, leaving room for additional units in the future if needed.

Reply:
Istio uses protobuf to define all its APIs. Here's a [duration field](https://github.com/istio/api/blob/f930ebfda37a76446f229bf7e465bbbf305501af/networking/v1beta1/virtual_service.proto#L618) example. In the generated GO binding that translates to [durationpb.Duration](https://pkg.go.dev/google.golang.org/protobuf/types/known/durationpb), similar API to `metav1.Duration`.

Reply:
OK, capturing some more context from the [Slack thread](https://kubernetes.slack.com/archives/CJUQN3E4T/p1688063522998959). @liggitt suggested that we could start with a simple regex that represented a small subset of these duration formats - something like this:

`^[1-9][0-9]*(h|m|s|ms)$`

As far as I can tell, that would be a subset of Go and XML/OpenAPI formats used here, likely many others as well. Importantly it would prevent specifying something like `1m30s` and instead you'd need to represent that as `90s`. Some potential modifications to make this more robust:

1. Allow a zero value to indicate "no timeout":

`^[0]|[1-9][0-9]*(h|m|s|ms)$`

2. Limit the number of digits to effectively enforce a max value:

`^[0]$|^[1-9][0-9]{0,3}(h|m|s|ms)$`

3. Remove hour from the allowed set of units since most timeouts can be expressed without them

`^[0]$|^[1-9][0-9]{0,3}(m|s|ms)$`

This would make the max value 9999 minutes which is ~167 hours or ~7 days.

Alternatively we could just choose a single value like milliseconds which would be easier to manage from both an API and implementation perspective, but decidedly more of a pain for users to configure.

Reply:
I've updated this PR to use the approach I've described in the comment above, PTAL.

/cc @frankbu @srodi @shaneutt @thockin @howardjohn

Reply:
If we do this, let's consider the precedent.  Hours seems required for anyone else who follows this pattern.

> As far as I can tell, that would be a subset of Go and XML/OpenAPI formats

I looked at the XML spec and I find 3 things that I can't answer for sure:

1) They define the units as "H", "M", "S", whereas go does "h", "m", "s".  I could not convince myself that case is ignored

2) They define all values as starting with literal "P"

3) They do not have a milliseconds unit.

As such, I think "compatible with XMLSchema" is not really true, and we're just defining a subset of Go's duration grammar.

Reply:
Regarding the 0 value, I would go a step further and regard a 0 timeout value as if the value was empty.  This indicates that the user doesn't care and will accept the default from the implementation.

Reply:
>Regarding the 0 value, I would go a step further and regard a 0 timeout value as if the value was empty.

@candita I think we need to avoid that here for backwards compatibility. We need the nil value to be implementation-specific and essentially map to whatever implementations are already doing. It's unfortunate to make 0 mean something different than nil, and I was initially against that, but it does seem like it's overwhelmingly used to represent "infinite" when it comes to timeout configuration in underlying implementations, so we should probably support that here as well.

Reply:
> Based on the feedback in yesterday's community meeting, I think the proposal itself of a simplified string format for duration is reasonable.

@liggitt are you OK with this?  It'll set a new precedent and impose client-side parsing.

If we do this, we should probably define a `type Duration string` and comment it with the full grammar.  We need to think about compat - if we don't allow `"1m30s"` or 'h' and 'd' suffixes or `"1.5m"` now, we may never be able to allow it (because client-side parsing).  Please don't take this lightly.

Reply:
> @liggitt are you OK with this? It'll set a new precedent and impose client-side parsing.

It doesn't seem worse than resource.Quantity or IntOrString we already have in other places.

> If we do this, we should probably define a type Duration string and comment it with the full grammar. We need to think about compat - if we don't allow "1m30s" or 'h' and 'd' suffixes or "1.5m" now, we may never be able to allow it (because client-side parsing). Please don't take this lightly.

As a general type, I'd definitely allow `h`, `0`, negative values, etc.

`d` is not defined in time.Duration, and "1 day" is not well-defined in the face of leap-seconds/daylight saving shifts/etc, so excluding that seems fine.

My instinct was to disallow fractional values, but maybe there's a case to be made for those?

Reply:
> It doesn't seem worse than resource.Quantity or IntOrString we already have in other places.

I have been vocal about NOT further propagating IntOrString (or recreating the same hackery).  Many people think quantity "could have been done better".  I don't think "not worse than" is a great criteria here.  I think we need to say it is affirmatively better than the alternatives.

1) `TimeoutSeconds int32`
2) `TimeoutMillseconds int64`
3) `Timeout struct { Seconds int32, Milliseconds int32 }`
4) `TimeoutSeconds int32; TimeoutMilliSecinds int32;`
5) `Timeout DurationString`

> As a general type, I'd definitely allow h, 0, negative values, etc.

Agree

> d is not defined in time.Duration, and "1 day" is not well-defined 

SGTM

> My instinct was to disallow fractional values...as simple and unambiguous as possible

My instinct, too.  I don't want to specify FP behavior and precision limits.  I don't want to impose fixed-point math on clients.  I see why *simple* FP is attractive but the grammar is sufficiently complicated that I think we can live without it.  I just want to think it through.

`/-?([0-9]+\.)?[1-9][0-9]*(h|m|s|ms)/` is not so bad, but I have 50% confidence that I actually got it right.  Do we actually support fractional milliseconds?  Do we limit it to 3 decimal places?

Reply:
We also need to NOT NORMALIZE it.

Reply:
While I'm as down on floats as the next person (🙂) , I feel like we should just go ahead and use `metav1.Duration` for the timeouts. It's already present and tested in apimachinery, it's starting to see use (KubeletConfiguration, for example), and it's well-supported in languages used in the ecosystem (Go is obvious, but a Rust parser is being added to [kube-rs](https://github.com/kube-rs/kube/pull/1224) as part of Buoyant's work implementing GEP-1742 timeouts in Linkerd, and there are implementations in Python as well).

Most importantly, though: metav1.Duration is easier for the user than other things. Sadly, my experience with Emissary-ingress is that having the unit in the name results in more errors than having it in the value -- I've seen far fewer cases of a user typoing `35s` when they meant `35ms` than I have of users e.g. putting a number of seconds into a field explicitly marked as milliseconds (e.g. `timeout_ms: 30` instead of `timeout_ms: 30000`). I think the Go folks did a nice job here of creating a thing that's simple for the user to use, even though it's harder on the implementers.

Extending @thockin's regex seems to work fine in quick tests:

```
^-?(([0-9]+\.)?[0-9]+(h|m|s|ms|us|ns))+$
```

We could simplify it if we want to disallow negative durations, floating-point values, or things smaller than milliseconds (I'd definitely be OK with the first two for timeouts). And we could expand it dramatically to catch out-of-order or repeating units, too, but I don't think that's worth it.

Reply:
> I feel like we should just go ahead and use `metav1.Duration` for the timeouts

I don't mind this idea, but for reference I can't find a way to both use this type and add the kind of regex validation that would represent a subset of `metav1.Duration`. Very helpful to see that work is already underway to support this format in rust and python though. Ideally I think we should just aim to have a string format that is a subset of `metav1.duration`.

>I think we need to say it is affirmatively better than the alternatives.

I believe it clearly provides a better UX than the alternatives. Although it may be messier for non-golang implementations, it is also consistent with how so many implementations already handle timeouts, including Contour, Istio, and Linkerd.

Would it be fair to separate the following?

1. The range of values that should be supported by this broader format (may include positive and negative values + `h|m|s|ms|us|ns` units)
2. The range of values that we want to support in Gateway API (a subset of 1)

While I agree that the concept of a duration should include more units and _maybe_ negative values, I really don't want to overcomplicate the values supported by Gateway API. In general we've tried to support just the most common formats that are going to be broadly implementable. Concepts like negative timeouts and/or microseconds or nanoseconds really don't make sense in this context, and I'd hate to let validation allow them through.

Reply:
I'm not convinced we need durations smaller than ms (famous last words, so I don't REALLY mind keeping `us` and `ns`, but not `µs`).

I think we can leave out unary `+` but keep unary `-`.

I think we can leave out compound values (e.g. no "1m30s").

I think we can leave out floating point values.  (my above regex was wrong.  `^-?([1-9][0-9]*|[0-9]*\.[0-9]{0,3})(h|m|s|ms)$` seems better, but let's not?)

That leaves us with something like `^-?[1-9][0-9]*(h|m|s|ms)$` - at least it's a) regexable and b) easy to parse by a human

Do we want to argue about FP or compound?

Reply:
>I think we can leave out unary + but keep unary -.

I would strongly prefer to leave that out from valid Gateway API values. Although we can define a broader type that allows negative values, I don't want to unnecessarily expand the set of values Gateway API supports. In the specific case of negative values, I don't think there's any possible reasonable interpretation for timeouts so I'd rather avoid them.

I think we need to differentiate between a broader standard that we want to introduce, the subset of that standard that we want to allow for this specific use case.

Reply:
One other variation that is critical for Gateway API but likely confusing for others. In most/all implementations I'm aware of, "0" is interpreted as no/infinite timeout. The variations of the regex I proposed allowed room for "0" without any units to communicate this state. If that's problematic, we could just tell users to configure something like "0s" to represent this concept.

Reply:
I don't mind allowing 0, but in YAML it will have to be `"0"` because the field will be transported as a string.  `0s` does not suffer that problem.

Reply:
>> I think we can leave out unary + but keep unary -.
>
> I would strongly prefer to leave that out from valid Gateway API values. Although we can define a broader type that allows negative values,

I'm talking about the difference between "Invalid value: timeout: must match regex ...." and "Invalid value: timeout: must be greater than or equal to 0ms"

Reply:
> I don't mind allowing 0, but in YAML it will have to be "0" because the field will be transported as a string. 0s does not suffer that problem.

Also not that comparing `0s` to `0ms` to `0h` must succeed, as must "1000ms" and `1s`

Reply:
>I'm talking about the difference between "Invalid value: timeout: must match regex ...." and "Invalid value: timeout: must be greater than or equal to 0ms"

Unfortunately I don't know of any way to do that with CRD validation. We could do that with webhook validation, but that is far less reliable so I'd rather cover as much as possible directly with validation defined on the CRD.

>Also not that comparing 0s to 0ms to 0h must succeed, as must "1000ms" and 1s

Agreed, we should likely have some conformance tests to confirm that.

Reply:
Catching up here, this is quite the rock to turn over!

I'm supportive of using a string based definition that's a subset of `metav1.Duration`, however we do it, and also agree with ensuring that we have conformance tests that validate as much behavior as possible.